presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -146,17 +150,30 @@ public int compare(PlanNode node1, PlanNode node2)
         }
 
         private PlanNode chooseJoinOrder(List<PlanNode> sources, List<Symbol> outputSymbols)
+                throws UnknownCostException
         {
             Set<PlanNode> multiJoinKey = ImmutableSet.copyOf(sources);
             PlanNode node = memo.get(multiJoinKey);
             if (node == null) {
                 checkState(sources.size() > 1);
-                node = generatePartitions(sources.size())
-                        .map(partitioning -> createJoinAccordingToPartitioning(sources, outputSymbols, partitioning))
-                        .filter(Optional::isPresent)
-                        .map(Optional::get)
-                        .min(planNodeOrdering)
-                        .orElseGet(() -> createCrossJoins(sources, outputSymbols));
+                try {
+                    node = generatePartitions(sources.size())
+                            .map(partitioning -> {
+                                try {
+                                    return createJoinAccordingToPartitioning(sources, outputSymbols, partitioning);
+                                }
+                                catch (UnknownCostException unknownCostException) {
+                                    throw new WrappedUnknownCostException(unknownCostException);
+                                }
+                            })
+                            .filter(Optional::isPresent)
+                            .map(Optional::get)
+                            .min(planNodeOrdering)
+                            .orElseGet(() -> createCrossJoins(sources, outputSymbols));
+                }
+                catch (WrappedUnknownCostException wrappedException) {
```

Comment:
this is needed because you can't throw checked exceptions within streams.  An alternative would be not to use a stream at all.  something like this:

```
Set<PlanNode> multiJoinKey = ImmutableSet.copyOf(sources);
            PlanNode bestPlan = memo.get(multiJoinKey);
            if (bestPlan == null) {
                checkState(sources.size() > 1);
                ImmutableList.Builder<PlanNode> plans = ImmutableList.builder();
                for (Set<Integer> partition : (Iterable<Set<Integer>>) generatePartitions(sources.size())::iterator) {
                    Optional<PlanNode> planNodeOptional = createJoinAccordingToPartitioning(sources, outputSymbols, partition);
                    if (planNodeOptional.isPresent()) {
                        PlanNode planNode = planNodeOptional.get();
                        PlanNodeCostEstimate cost = lookup.getCumulativeCost(planNode, session, symbolAllocator.getTypes());
                        if (cost.isUnknown()) {
                            throw new UnknownCostException();
                        }
                        plans.add(planNode);
                    }
                }

                bestPlan = plans.build().stream()
                        .min(planNodeOrdering)
                        .orElseGet(() -> createCrossJoins(sources, outputSymbols));
                log.debug("Least cost join was: " + bestPlan.toString());
                memo.put(multiJoinKey, bestPlan);
            }
            return bestPlan;
```